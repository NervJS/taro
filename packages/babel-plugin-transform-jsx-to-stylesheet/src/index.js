const path = require('path')

const STYLE_SHEET_NAME = '_styleSheet'
const GET_STYLE_FUNC_NAME = '_getStyle'
const MERGE_STYLES_FUNC_NAME = '_mergeStyles'
const GET_CLS_NAME_FUNC_NAME = '_getClassName'
const NAME_SUFFIX = 'StyleSheet'
const cssSuffixs = ['.css', '.scss', '.sass', '.less', '.styl']
let styleNames = [] // css module 写法中 import 的 Identifier

module.exports = function ({types: t, template}) {
  const mergeStylesFunctionTemplate = template(`
function ${MERGE_STYLES_FUNC_NAME}() {
  var newTarget = {};

  for (var index = 0; index < arguments.length; index++) {
    var target = arguments[index];

    for (var key in target) {
      newTarget[key] = Object.assign(newTarget[key] || {}, target[key]);
    }
  }

  return newTarget;
}
  `)
  const getClassNameFunctionTemplate = template(`
function ${GET_CLS_NAME_FUNC_NAME}() {
  var className = [];
  var args = arguments[0];
  var type = Object.prototype.toString.call(args).slice(8, -1).toLowerCase();

  if (type === 'string') {
    args = args.trim();
    args && className.push(args);
  } else if (type === 'array') {
    args.forEach(function (cls) {
      cls = ${GET_CLS_NAME_FUNC_NAME}(cls).trim();
      cls && className.push(cls);
    });
  } else if (type === 'object') {
    for (var k in args) {
      k = k.trim();
      if (k && args.hasOwnProperty(k) && args[k]) {
        className.push(k);
      }
    }
  }

  return className.join(' ').trim();
}
  `)
  const getStyleFunctionTemplete = template(`
function ${GET_STYLE_FUNC_NAME}(classNameExpression) { 
  var className = ${GET_CLS_NAME_FUNC_NAME}(classNameExpression);
  var classNameArr = className.split(/\\s+/);

  var style = [];
  if (classNameArr.length === 1) {
    style.push(${STYLE_SHEET_NAME}[classNameArr[0].trim()]);
  } else {
      classNameArr.forEach(function(cls) {
      style.push(${STYLE_SHEET_NAME}[cls.trim()]);
    });
  }

  return style;
}
  `)

  const getClassNameFunctionAst = getClassNameFunctionTemplate()
  const mergeStylesFunctionAst = mergeStylesFunctionTemplate()
  const getStyleFunctionAst = getStyleFunctionTemplete()

  function getArrayExpression (value) {
    let expression
    let str

    if (!value || value.value === '') {
      // className
      // className=""
      return []
    } else if (value.type === 'JSXExpressionContainer' && value.expression && typeof value.expression.value !== 'string') {
      // className={{ container: true }}
      // className={['container wrapper', { scroll: false }]}
      return [t.callExpression(t.identifier(GET_STYLE_FUNC_NAME), [value.expression])]
    } else {
      // className="container"
      // className={'container'}
      str = (value.expression ? value.expression.value : value.value).trim()
    }

    return str === '' ? [] : str.split(/\s+/).map((className) => {
      return template(`${STYLE_SHEET_NAME}["${className}"]`)().expression
    })
  }

  function findLastImportIndex (body) {
    const bodyReverse = body.slice(0).reverse()
    let _index = 0

    bodyReverse.some((node, index) => {
      if (node.type === 'ImportDeclaration') {
        _index = body.length - index - 1
        return true
      }
      return false
    })

    return _index
  }

  return {
    visitor: {
      Program: {
        exit (astPath, {file}) {
          const node = astPath.node
          // const cssFileCount = file.get('cssFileCount')
          const injectGetStyle = file.get('injectGetStyle')
          const lastImportIndex = findLastImportIndex(node.body)
          let cssParamIdentifiers = file.get('cssParamIdentifiers')
          let callExpression

          if (cssParamIdentifiers) {
            // only one css file，由于样式文件合并，永远只有一个
            if (cssParamIdentifiers.length === 1) {
              callExpression = t.variableDeclaration('var', [t.variableDeclarator(t.identifier(STYLE_SHEET_NAME), cssParamIdentifiers[0])])
            } else if (cssParamIdentifiers.length > 1) {
              const objectAssignExpression = t.callExpression(t.identifier(MERGE_STYLES_FUNC_NAME), cssParamIdentifiers)
              callExpression = t.variableDeclaration('var', [t.variableDeclarator(t.identifier(STYLE_SHEET_NAME), objectAssignExpression)])
            }

            node.body.splice(lastImportIndex + 1, 0, callExpression)

            if (injectGetStyle) {
              node.body.splice(lastImportIndex + 2, 0, getClassNameFunctionAst)
              node.body.splice(lastImportIndex + 3, 0, getStyleFunctionAst)
            }
          }

          // 全部合并成一个，也就不需要 mergeStylesFunctionAst 了
          // if (cssFileCount > 1) {
          //   node.body.unshift(mergeStylesFunctionAst)
          // }

          // 将 css modules 的 styles 替换为 _styleSheet
          astPath.traverse({
            JSXOpeningElement (astPath) {
              astPath.traverse({
                JSXAttribute ({node}) {
                  if (node.name && node.name.name === 'style') {
                    astPath.traverse({
                      MemberExpression ({node}) {
                        if (node.object.type === 'Identifier' && styleNames.indexOf(node.object.name) > -1) {
                          node.object.name = STYLE_SHEET_NAME
                        }
                      }
                    })
                  }
                }
              })
            }
          })
          // 清除该 js 文件的 styleNames
          styleNames = []
        }
      },
      JSXOpeningElement ({container}, {file}) {
        const cssFileCount = file.get('cssFileCount') || 0
        if (cssFileCount < 1) {
          return
        }

        // Check if has "style"
        let hasStyleAttribute = false
        let styleAttribute
        let hasClassName = false
        let classNameAttribute

        const attributes = container.openingElement.attributes
        for (let i = 0; i < attributes.length; i++) {
          const name = attributes[i].name
          if (name) {
            if (!hasStyleAttribute) {
              hasStyleAttribute = name.name === 'style'
              styleAttribute = hasStyleAttribute && attributes[i]
            }

            if (!hasClassName) {
              hasClassName = name.name === 'className'
              classNameAttribute = hasClassName && attributes[i]
            }
          }
        }

        if (hasClassName) {
          // Remove origin className
          attributes.splice(attributes.indexOf(classNameAttribute), 1)

          if (
            classNameAttribute.value &&
            classNameAttribute.value.type === 'JSXExpressionContainer' &&
            typeof classNameAttribute.value.expression.value !== 'string' // not like className={'container'}
          ) {
            file.set('injectGetStyle', true)
          }

          const arrayExpression = getArrayExpression(classNameAttribute.value)

          if (arrayExpression.length === 0) {
            return
          }

          if (hasStyleAttribute && styleAttribute.value) {
            let expression = styleAttribute.value.expression
            let expressionType = expression.type

            // style={[styles.a, styles.b]} ArrayExpression
            if (expressionType === 'ArrayExpression') {
              expression.elements = arrayExpression.concat(expression.elements)
              // style={styles.a} MemberExpression
              // style={{ height: 100 }} ObjectExpression
              // style={{ ...custom }} ObjectExpression
              // style={custom} Identifier
              // style={getStyle()} CallExpression
              // style={this.props.useCustom ? custom : null} ConditionalExpression
              // style={custom || other} LogicalExpression
            } else {
              styleAttribute.value.expression = t.arrayExpression(arrayExpression.concat(expression))
            }
          } else {
            let expression = arrayExpression.length === 1 ? arrayExpression[0] : t.arrayExpression(arrayExpression)
            attributes.push(t.jSXAttribute(t.jSXIdentifier('style'), t.jSXExpressionContainer(expression)))
          }
        }
      },
      // 由于目前 js 引入的文件样式默认会全部合并，故进插入一个就好，其余的全部 remove
      ImportDeclaration (astPath, state) {
        const {file} = state
        const node = astPath.node
        const sourceValue = node.source.value
        const specifiers = node.specifiers
        const jsFilePath = state.opts.filePath // 传进来的文件的 filePath ,Babel6 file.opts.filaname 为unknown
        const extname = path.extname(sourceValue)
        const cssIndex = cssSuffixs.indexOf(extname)
        let cssFileCount = file.get('cssFileCount') || 0
        let cssParamIdentifiers = file.get('cssParamIdentifiers') || []

        if (cssIndex > -1) {
          // `import styles from './foo.css'` kind
          if (node.importKind === 'value' && specifiers.length > 0) {
            styleNames.push(specifiers[0].local.name)
          }
          // 第一个引入的样式文件
          if (cssFileCount === 0) {
            const cssFileBaseName = path.basename(jsFilePath, path.extname(jsFilePath))
            // 引入样式对应的变量名
            const styleSheetIdentifierValue = `${cssFileBaseName.replace(/-/g, '_') + NAME_SUFFIX}`
            const styleSheetIdentifierPath = `./${cssFileBaseName}_styles`
            const styleSheetIdentifier = t.identifier(styleSheetIdentifierValue)

            node.specifiers = [t.importDefaultSpecifier(styleSheetIdentifier)]
            node.source = t.stringLiteral(styleSheetIdentifierPath)
            cssParamIdentifiers.push(styleSheetIdentifier)
          } else {
            astPath.remove()
          }
          cssFileCount++

          file.set('cssParamIdentifiers', cssParamIdentifiers)
          file.set('cssFileCount', cssFileCount)
        }
      }
    }
  }
}
