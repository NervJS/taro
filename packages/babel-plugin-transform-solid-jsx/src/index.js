import SyntaxJSX from '@babel/plugin-syntax-jsx'

import postprocess from './shared/postprocess'
import preprocess from './shared/preprocess'
import { transformJSX } from './shared/transform'

export default () => {
  return {
    name: 'JSX DOM Expressions',
    inherits: SyntaxJSX.default,
    visitor: {
      JSXElement: transformJSX,
      JSXFragment: transformJSX,
      Program: {
        enter: preprocess,
        exit: postprocess
      }
    }
  }
}
