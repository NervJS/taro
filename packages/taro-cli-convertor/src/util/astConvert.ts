import generate from '@babel/generator'
import { NodePath } from '@babel/traverse'
import * as t from '@babel/types'

// 最低限度的转义： https://github.com/mathiasbynens/jsesc#minimal
export function generateMinimalEscapeCode (ast: t.File) {
  return generate(ast as any, {
    jsescOption: {
      minimal: true,
    },
  }).code
}

// 判断是否已经引入 @tarojs/taro
export function hasTaroImport (bodyNode: NodePath<t.Node>[]) {
  return bodyNode.some((astPath) => {
    // import Taro from '@tarojs/taro'
    if (astPath.isImportDeclaration()) {
      return astPath.node.source.value === '@tarojs/taro'
    } else if (astPath.isVariableDeclaration() && astPath.node.declarations.length === 1) {
      // const Taro = require('@tarojs/taro')
      const declaration: t.VariableDeclarator = astPath.node.declarations[0]
      return (
        declaration.init != null &&
        declaration.init.type === 'CallExpression' &&
        declaration.init.callee.type === 'Identifier' &&
        declaration.init.callee.name === 'require' &&
        declaration.init.arguments.length === 1 &&
        declaration.init.arguments[0].type === 'StringLiteral' &&
        declaration.init.arguments[0].value === '@tarojs/taro'
      )
    }
    return false
  })
}

// 根据关键字exports判断是否为commonjs模块
export function isCommonjsModule (bodyNode: NodePath<t.Node>[]) {
  return bodyNode.some((p) => {
    if (t.isExpressionStatement(p) && t.isAssignmentExpression(p.expression)) {
      const expression = p.expression
      // 1、module.exports.num = num 2、module.exports = {}
      const isModuleExports =
        (expression.left.type === 'MemberExpression' &&
          expression.left.object.type === 'MemberExpression' &&
          expression.left.object.property.type === 'Identifier' &&
          expression.left.object.property.name === 'exports') ||
        (expression.left.type === 'MemberExpression' &&
          expression.left.property.type === 'Identifier' &&
          expression.left.property.name === 'exports')

      // exports.num = num
      const isExports =
        expression.left.type === 'MemberExpression' &&
        expression.left.object.type === 'Identifier' &&
        expression.left.object.name === 'exports'
      return isModuleExports || isExports
    }
    return false
  })
}

/**
 * 判断节点是否为commonjs的导入
 *
 * @param {NodePath<t.Node>} bodyNode
 * @returns {boolean}
 */
export function isCommonjsImport (bodyNode: NodePath<t.Node>) {
  if (bodyNode.isVariableDeclaration() && bodyNode.node.declarations.length === 1) {
    const declaration: t.VariableDeclarator = bodyNode.node.declarations[0]
    return (
      declaration.init != null &&
      declaration.init.type === 'CallExpression' &&
      declaration.init.callee.type === 'Identifier' &&
      declaration.init.callee.name === 'require'
    )
  }
  return false
}
