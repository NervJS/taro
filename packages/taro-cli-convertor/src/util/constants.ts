export const enum Constants {
  PLUGIN = 'plugin',

  PLUGIN_JSON = 'plugin.json',
}
