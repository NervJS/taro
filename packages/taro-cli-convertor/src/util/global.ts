import { ErrCodeMsg } from './index'

// taro-cli-convertor 模块的全局变量
interface Globals {
  logFilePath: string   // 日志文件路径
  errCodeMsgs: ErrCodeMsg[]   // 错误代码信息 (转换报告涉及的字段)
  currentParseFile: string    // 当前解析的文件 (转换报告涉及的字段)
  logCount: number 
  logFileContent: string
}

export const globals: Globals = {
  logFilePath: '',
  errCodeMsgs: [],
  currentParseFile: '',
  logCount: 0,
  logFileContent: ''
}
