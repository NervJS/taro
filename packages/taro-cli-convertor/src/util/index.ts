import {
  chalk,
  CSS_IMPORT_REG,
  fs,
  printLog,
  processTypeEnum,
  promoteRelativePath,
  REG_SCRIPT,
  REG_TYPESCRIPT,
  resolveScriptPath,
  SCRIPT_EXT,
} from '@tarojs/helper'
import * as path from 'path'

import type * as t from '@babel/types'

const NODE_MODULES = 'node_modules'

interface IReportMsg {
  filePath: string // 报告信息文件路径
  message: string // 报告信息
  type?: string // 报告信息类型
  childReportMsg?: IReportMsg[]
}

export function getRootPath (): string {
  return path.resolve(__dirname, '../../')
}

export function getPkgVersion (): string {
  return require(path.join(getRootPath(), 'package.json')).version
}

// 文件存在或添加后缀.js、.jsx、.ts、.tsx存在则返回文件路径，否则返回null
function revertScriptPath (absolutePath: string, SCRIPT_EXT: string[]) {
  for (const item of SCRIPT_EXT) {
    if (fs.existsSync(absolutePath)) {
      return absolutePath
    } else if (fs.existsSync(`${absolutePath}${item}`)) {
      return `${absolutePath}${item}`
    }
  }
  return null
}

function getRelativePath (_rootPath: string, sourceFilePath: string, oriPath: string) {
  // 处理以/开头的绝对路径，比如 /a/b
  if (path.isAbsolute(oriPath)) {
    if (oriPath.indexOf('/') !== 0) {
      return ''
    }
    const absolutePath = revertScriptPath(path.resolve(sourceFilePath, '..' + oriPath), SCRIPT_EXT)
    if (absolutePath == null) {
      return ''
    }

    let relativePath = path.relative(path.dirname(sourceFilePath), absolutePath)
    relativePath = promoteRelativePath(relativePath)
    if (relativePath.indexOf('.') !== 0) {
      return `./${relativePath}`
    }
    return relativePath
  }
  // 处理非正常路径，比如 a/b
  if (oriPath.indexOf('.') !== 0) {
    const absolutePath = revertScriptPath(path.resolve(sourceFilePath, '..', oriPath), SCRIPT_EXT)

    // 可能为三方库
    if (absolutePath == null) {
      return oriPath
    }

    let relativePath = path.relative(path.dirname(sourceFilePath), absolutePath)
    relativePath = promoteRelativePath(relativePath)
    if (relativePath.indexOf('.') !== 0) {
      return `./${relativePath}`
    }
    return relativePath
  }
  return oriPath
}

export function copyFileToTaro (from: string, to: string, options?: fs.CopyOptionsSync) {
  const filename = path.basename(from)
  if (fs.statSync(from).isFile() && !path.extname(to)) {
    fs.ensureDir(to)
    return fs.copySync(from, path.join(to, filename), options)
  }
  fs.ensureDir(path.dirname(to))
  return fs.copySync(from, to, options)
}

// 递归复制文件夹
function copyFolderToTaro (source: string, destination: string) {
  // 创建目标文件夹
  if (!fs.existsSync(destination)) {
    fs.mkdirSync(destination, { recursive: true })
  }

  // 获取源文件夹中的所有项
  const items = fs.readdirSync(source)
  for (const item of items) {
    const sourcePath = path.join(source, item)
    const destinationPath = path.join(destination, item)

    const stat = fs.lstatSync(sourcePath)
    if (stat.isSymbolicLink()) {
      // 如果是符号链接，获取链接的目标路径并进行复制
      const target = fs.readlinkSync(sourcePath)
      copyFolderToTaro(target, destinationPath)
    } else if (stat.isDirectory()) {
      // 如果是子文件夹，则递归复制
      copyFolderToTaro(sourcePath, destinationPath)
    } else {
      // 如果是普通文件，则直接复制
      fs.copyFileSync(sourcePath, destinationPath)
    }
  }
}

export function analyzeImportUrl (
  rootPath: string,
  sourceFilePath: string,
  scriptFiles: Set<string>,
  source: t.StringLiteral,
  value: string,
  isTsProject?: boolean
) {
  const valueExtname = path.extname(value)
  const rpath = getRelativePath(rootPath, sourceFilePath, value)
  if (!rpath) {
    printLog(processTypeEnum.ERROR, '引用文件', `文件 ${sourceFilePath} 中引用 ${value} 不存在！`)
    return
  }
  if (rpath !== value) {
    value = rpath
    source.value = rpath
  }
  if (value.indexOf('.') === 0) {
    if (REG_SCRIPT.test(valueExtname) || REG_TYPESCRIPT.test(valueExtname)) {
      const vpath = path.resolve(sourceFilePath, '..', value)
      let fPath = value
      if (fs.existsSync(vpath)) {
        fPath = vpath
      } else {
        printLog(processTypeEnum.ERROR, '引用文件', `文件 ${sourceFilePath} 中引用 ${value} 不存在！`)
      }
      scriptFiles.add(fPath)
    } else {
      let vpath = resolveScriptPath(path.resolve(sourceFilePath, '..', value))
      if (vpath) {
        if (!fs.existsSync(vpath)) {
          printLog(processTypeEnum.ERROR, '引用文件', `文件 ${sourceFilePath} 中引用 ${value} 不存在！`)
        } else {
          if (fs.lstatSync(vpath).isDirectory()) {
            if (fs.existsSync(path.join(vpath, 'index.js'))) {
              vpath = path.join(vpath, 'index.js')
            } else {
              printLog(processTypeEnum.ERROR, '引用目录', `文件 ${sourceFilePath} 中引用了目录 ${value}！`)
              return
            }
          }
          let relativePath = path.relative(sourceFilePath, vpath)
          const relativePathExtname = path.extname(relativePath)
          scriptFiles.add(vpath)
          relativePath = promoteRelativePath(relativePath)
          if (/\.wxs/.test(relativePathExtname)) {
            relativePath += '.js'
          } else if (!isTsProject) {
            relativePath = relativePath.replace(relativePathExtname, '.js')
          }
          source.value = relativePath
        }
      }
    }
  } else {
    if (value.startsWith('/') || value.startsWith('@tarojs') || value.startsWith('react')) {
      return
    }
    scriptFiles.add(value)
  }
}

export const incrementId = () => {
  let n = 0
  return () => (n++).toString()
}

// 处理三方库引用
export function handleThirdPartyLib (filePath: string, nodePath: string[], root: string, convertRoot: string) {
  // 默认使用node_modules中的三方库
  if (typeof nodePath === 'undefined') {
    nodePath = [NODE_MODULES]
  }

  try {
    let isThirdPartyLibExist = false
    for (const modulePath of nodePath) {
      const parts = filePath.split('/')
      const npmModulePath = path.resolve(root, modulePath, parts[0])
      if (fs.existsSync(npmModulePath)) {
        isThirdPartyLibExist = true
        // 转换后的三方库放在node_modules中
        const moduleConvertPath = path.resolve(convertRoot, NODE_MODULES, parts[0])
        if (!fs.existsSync(moduleConvertPath)) {
          copyFolderToTaro(npmModulePath, moduleConvertPath)
          printLog(processTypeEnum.COPY, '三方库', normalizePath(npmModulePath.replace(path.join(root, '/'), '')))
        }
        break
      }
    }

    if (!isThirdPartyLibExist) {
      console.log(chalk.red(`在[${nodePath.toString()}]中没有找到依赖的三方库${filePath}，请安装依赖后运行`))
    }
  } catch (error) {
    console.log(chalk.red(`转换三方库${filePath}异常，请手动处理, error message:${error}`))
  }
}

export function getMatchUnconvertDir (importPath: string, externalPaths: string[]) {
  if (importPath === undefined || externalPaths === undefined) {
    return null
  }
  // 支持用户在convert.config.json中配置不转换的目录
  for (let externalPath of externalPaths) {
    externalPath = normalizePath(externalPath).replace('*', '.*')
    const reg = new RegExp(externalPath)
    const match = reg.exec(normalizePath(importPath))
    // external字段配置如 1：../demo 2：../demo/* 3：../demo/*/demo
    if (match) {
      // 处理../demo/*
      if (externalPath.length >= 2 && externalPath.endsWith('*')) {
        // 去掉结尾的".*"
        return externalPath.slice(0, externalPath.length - 3)
      }
      return match[0]
    }
  }
  return null
}

export function handleUnconvertDir (matchUnconvertDir: string, rootPath: string, convertRoot: string) {
  if (matchUnconvertDir == null) {
    return
  }
  // 支持用户在convert.config.json中配置不转换的目录
  const outputFilePath = matchUnconvertDir.replace(normalizePath(rootPath), normalizePath(convertRoot))
  if (!fs.existsSync(outputFilePath)) {
    copyFileToTaro(matchUnconvertDir, outputFilePath)
    printLog(processTypeEnum.COPY, '不转换目录', matchUnconvertDir.replace(normalizePath(path.join(rootPath, '/')), ''))
  }
}

export function getMatchDirPath (filePath: string, matchStr: string) {
  const reg = new RegExp(matchStr)
  const match = reg.exec(filePath)
  if (match) {
    if (matchStr.length >= 2 && matchStr.endsWith('*')) {
      return matchStr.slice(0, matchStr.length - 2)
    }
    return match[0]
  }
}

// 路径标准化
function normalizePath (path) {
  if (typeof path === 'undefined') {
    return ''
  }
  return path.replace(/\\/g, '/')
}

// 将数组中的相对路径转换为绝对路径
export function transRelToAbsPath (baseFilePath: string, fileRelativePaths: string[]) {
  if (typeof fileRelativePaths === 'undefined') {
    return []
  }
  const absolutePath: string[] = []
  for (const fileRelativePath of fileRelativePaths) {
    // 相对路径转为绝对路径
    absolutePath.push(path.resolve(baseFilePath, '..', fileRelativePath))
  }
  return absolutePath
}

// 获取wxss中引用文件的相对路径集合
export function getWxssImports (content: string) {
  if (content == null) {
    return []
  }
  // 匹配 /* ... */ 形式的注释
  const regex = /\/\*([\s\S]*?)\*\//g
  // 去掉css中的注释
  const contentWithoutComment = content.replace(regex, '')

  let match
  const imports: string[] = []
  const cssImportReg = new RegExp(CSS_IMPORT_REG)
  while ((match = cssImportReg.exec(contentWithoutComment)) !== null) {
    imports.push(match[2])
  }
  return imports
}

/**
 * copyFileTo： 将报告模版文件复制到转换后 taroConvert 目录中
 *
 * @param { string } sourceFilePath 源文件路径
 * @param { string } targeFileDir 转换后文件所在目录
 * @param { string } targeFileName 转换后文件名
 * @param { IReportMsg[] } reportErroMsg 报错信息
 */
export function generateReportFile (sourceFilePath, targeFileDir, targeFileName, reportErroMsg?: IReportMsg[]) {
  try {
    if (!fs.existsSync(targeFileDir)) {
      fs.mkdirSync(targeFileDir)
    }
    let data = fs.readFileSync(sourceFilePath, 'utf-8')
    if (reportErroMsg) {
      data = data.replace('__errorMsgReport__', JSON.stringify(reportErroMsg))
    }
    fs.writeFileSync(path.join(targeFileDir, targeFileName), data)
  } catch (error) {
    console.log(`文件${sourceFilePath}写入失败，errorMsg：${error}`)
  }
}

// eslint-disable-next-line camelcase
export const DEFAULT_Component_SET = new Set<string>([
  'View',
  'Icon',
  'Progress',
  'RichText',
  'Text',
  'Button',
  'Checkbox',
  'CheckboxGroup',
  'Form',
  'Input',
  'Label',
  'Picker',
  'PickerView',
  'PickerViewColumn',
  'Radio',
  'RadioGroup',
  'Slider',
  'Switch',
  'CoverImage',
  'Textarea',
  'CoverView',
  'MovableArea',
  'MovableView',
  'ScrollView',
  'Swiper',
  'SwiperItem',
  'Navigator',
  'Audio',
  'Camera',
  'Image',
  'LivePlayer',
  'Video',
  'Canvas',
  'Ad',
  'WebView',
  'Block',
  'Map',
  'Slot',
  'SlotView',
  'Editor',
  'MatchMedia',
  'FunctionalPageNavigator',
  'LivePusher',
  'OfficialAccount',
  'OpenData',
  'NavigationBar',
  'PageMeta',
  'VoipRoom',
  'AdCustom',
])
