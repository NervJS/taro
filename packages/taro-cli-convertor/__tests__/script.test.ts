import * as taroize from '@tarojs/taroize'

import Convertor from '../src/index'
import { copyFileToTaro, getMatchUnconvertDir, transRelToAbsPath } from '../src/util'

const fs = require('fs')
const path = require('path')

describe('语法转换', () => {
  test('使用新建的setData替换组件中this.data.xx，实现this.data.xx的转换', () => {
    const entryJSON = { 'pages': ['pages/index/index'] }
    // json：index.json的内容  path：index的根目录（文件路径）  rootPath：小程序的根目录（文件路径）   
    // script：index.js的内容   scriptPath：index.js的绝对路径  wxml：index.html的内容
    const param = {
      json:'{}',
      path:'',
      rootPath:'',
      script:'this.data.intData = 1024',
      scriptPath:'',
      wxml:''
    }
    // new Convertot后会直接执行 init()，为确保 init() 在测试中通过采用 spyOn 去模拟
    jest.spyOn(Convertor.prototype, 'getApp').mockImplementation(() => {
      Convertor.prototype.entryJSON = entryJSON
    })
    jest.spyOn(Convertor.prototype, 'getPages').mockImplementation(() => {
      Convertor.prototype.pages = new Set(entryJSON.pages)
    })
    const convert = new Convertor('', false)
    convert.pages = Convertor.prototype.pages
    const taroizeResult = taroize({
      ...param,
      framework:'react'
    })
    // sourceFilePath：需要转换的文件路径   outputFilePath：转换输出路径   importStylePath：style的文件路径
    const { ast } = convert.parseAst({
      ast:taroizeResult.ast,
      sourceFilePath:'',
      outputFilePath:'',
      importStylePath:'',
      depComponents: new Set(),
      imports:[]
    })
    expect(ast).toMatchSnapshot()
  })

  test('组件的动态名称转换', () => {
    const entryJSON = { 'pages': ['pages/index/index'] }
    // json：index.json的内容  path：index的根目录（文件路径）  rootPath：小程序的根目录（文件路径）   
    // script：index.js的内容   scriptPath：index.js的绝对路径  wxml：index.html的内容
    const param = {
      json:'{}',
      path:'',
      rootPath:'',
      script:'',
      scriptPath:'',
      wxml:`<view wx:for="{{infoList}}" wx:key="infoId">
              <template is="info-{{item.tempName}}" data="{{item}}"></template>
            </view>`
    }
    // new Convertot后会直接执行 init()，为确保 init() 在测试中通过采用 spyOn 去模拟
    jest.spyOn(Convertor.prototype, 'getApp').mockImplementation(() => {
      Convertor.prototype.entryJSON = entryJSON
    })
    jest.spyOn(Convertor.prototype, 'getPages').mockImplementation(() => {
      Convertor.prototype.pages = new Set(entryJSON.pages)
    })
    const convert = new Convertor('', false)
    convert.pages = Convertor.prototype.pages
    const taroizeResult = taroize({
      ...param,
      framework:'react'
    })
    // sourceFilePath：需要转换的文件路径   outputFilePath：转换输出路径   importStylePath：style的文件路径
    const { ast } = convert.parseAst({
      ast:taroizeResult.ast,
      sourceFilePath:'',
      outputFilePath:'',
      importStylePath:'',
      depComponents: new Set(),
      imports:[]
    })
    expect(ast).toMatchSnapshot()
  })

  test('css中字母+数字+pX会转换成px', async () => {
    const convert = new Convertor('', false)
    const { css } = await convert.styleUnitTransform('', 'background-image: url("data:image/png;base64,TB0pX/TB0PX/TB0rpX/TB0RPX");')
    expect(css).toBe('background-image: url("data:image/png;base64,TB0pX/TB0PX/TB0rpX/TB0RPX");')
  })
})

describe('文件转换', () => {
  test('拷贝tsconfig.json文件到转换后的工程', () => {
    const convert = new Convertor('', false)
    const selfDefinedConfig: any = []
    // 目前只有tsconfig.json，还有的话继续加到array里
    selfDefinedConfig[0] = `tsconfig${convert.fileTypes.CONFIG}`
    for (const tempConfig of selfDefinedConfig) {
      const tempConfigPath = path.join(convert.root, tempConfig)
      if (fs.existsSync(tempConfig)) {
        const outputFilePath = path.join(convert.convertRoot, tempConfig)
        copyFileToTaro(tempConfigPath, outputFilePath)
        expect(fs.existsSync(outputFilePath)).toBe(true)
      }
    }
  })

  test('适配convert.config.json，对符合json内路径的文件，在其被导入时不做转换', () => {
    const convert = new Convertor('', false)
    const root = transRelToAbsPath('', ['./taro-cli-convertor'])[0].replace(/\\/g, '/')
    // 处理convert.config.json，并存储到convertConfig中
    const convertJsonPath: string = path.join(root, `convert.config${convert.fileTypes.CONFIG}`)
    const convertJson = { 'external': ['./taroConvert/'] }
    const convertConfig = { ...convertJson }
    convertConfig.external = transRelToAbsPath(convertJsonPath, convertConfig.external)
    // 处理不转换的目录，可在convert.config.json中external字段配置
    const filePath = transRelToAbsPath(convertJsonPath, ['./taroConvert/tsconfig.json'])
    const matchUnconvertDir: string | null = getMatchUnconvertDir(filePath[0], convertConfig?.external)
    if (matchUnconvertDir !== null) {
      // 支持用户在convert.config.json中配置不转换的目录
      const outputFilePath = 'src/taroConvert'
      if (!fs.existsSync(outputFilePath)) {
        copyFileToTaro(matchUnconvertDir, outputFilePath)
      }
      expect(fs.existsSync(outputFilePath)).toBe(true)
    }
  })
})