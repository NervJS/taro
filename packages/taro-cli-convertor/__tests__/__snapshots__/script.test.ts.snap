// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`page页面转换 template组件名转换 1`] = `
import withWeapp, { cacheOptions } from "@tarojs/with-weapp";
import { Block } from "@tarojs/components";
import React from "react";
import Taro from "@tarojs/taro";
import AnFishTmpl from 'taroConvert/src/imports/AnFishTmpl.js';
import ATmpl from 'taroConvert/src/imports/ATmpl.js';
import AshMerTmpl from 'taroConvert/src/imports/AshMerTmpl.js';
import AolTmpl from 'taroConvert/src/imports/AolTmpl.js';
import AshTmpl from 'taroConvert/src/imports/AshTmpl.js';
cacheOptions.setOptionsToCache({});
@withWeapp(cacheOptions.getOptionsFromCache())
class _C extends React.Component {
  render() {
    return <Block><AshTmpl></AshTmpl>{
        /*  template的name值为首字母大写且单个单词  */
      }<AolTmpl></AolTmpl>{
        /*  template的name值为首字母大写且多个单词  */
      }<AshMerTmpl></AshMerTmpl>{
        /*  template的name值为单字母  */
      }<ATmpl></ATmpl>{
        /*  template的name值为小驼峰  */
      }<AnFishTmpl></AnFishTmpl></Block>;
  }
}
export default _C;
`;

exports[`语法转换 使用新建的setData替换组件中this.data.xx，实现this.data.xx的转换 1`] = `
import withWeapp, { cacheOptions } from "@tarojs/with-weapp";
import { Block } from "@tarojs/components";
import React from "react";
import Taro from "@tarojs/taro";
this.setData({
  intData: 1024
});
cacheOptions.setOptionsToCache({});
@withWeapp(cacheOptions.getOptionsFromCache())
class _C extends React.Component {
  render() {
    return null;
  }
}
export default _C;
`;

exports[`语法转换 模版的动态名称转换 1`] = `
import withWeapp, { cacheOptions } from "@tarojs/with-weapp";
import { Block, View } from "@tarojs/components";
import React from "react";
import Taro from "@tarojs/taro";
this.setData({
  intData: 1024
});
cacheOptions.setOptionsToCache({});
const ComponentMap = {};
@withWeapp(cacheOptions.getOptionsFromCache())
class _C extends React.Component {
  render() {
    const {
      infoList
    } = this.data;
    return infoList.map((item, index) => {
      let ComponentName = ComponentMap["info-" + item.tempName];
      return <View key={item.infoId}><ComponentName data={{
          item: item
        }}></ComponentName></View>;
    });
  }
}
export default _C;
`;
